synapse/handlers/federation.py

```diff
@@ -1106,10 +1191,19 @@ async def _maybe_backfill_inner(
             redact=False,
             check_history_visibility_only=True,
         )
+        logger.info(
+            "_maybe_backfill_inner: filtered_extremities %s", filtered_extremities
+        )
 
-        if not filtered_extremities:
+        if not filtered_extremities and not insertion_events_to_be_backfilled:
             return False
 
+        extremities = {
+            **oldest_events_with_depth,
+            # TODO: insertion_events_to_be_backfilled is currently skipping the filtered_extremities checks
```

Comment:
Does it even matter that `insertion_events_to_be_backfilled` is not going through the same `filtered_extremities` checks?

We don't even use `filtered_extremities` after we calculate it. Maybe that is a bug though 🤔. Currently, we filter the `extremities`, check if there is any left, and if there is, go back to using the full `extremities` in the rest of the code.

Reply:
I'm open to filtering the `insertion_events_to_be_backfilled`, just give the nudge ⏩

Reply:
I think the logic behind using `extremities` rather than `filtered_extremities` is that if we are making a `/backfill` request we may as well send over all the information we have. We're not really trying to filter the extremities but instead trying to answer the question of "should we even bother calling `/backfill`? There's an argument that it might be better to only use the filtered extremities, but ideally the remote will do the right thing.

> I'm open to filtering the `insertion_events_to_be_backfilled`, just give the nudge

I think it probably doesn't make sense to filter it, as the logic of checking the history visibility doesn't really apply? The worst case here is that we backfill, get a bunch of redacted events, and then next time we hit this code path we'll likely not have an insertion event and so the filtering logic will work.